packages/block-editor/src/components/block-list-appender/index.js

```diff
@@ -8,56 +8,61 @@ import { last } from 'lodash';
  */
 import { withSelect } from '@wordpress/data';
 import { getDefaultBlockName } from '@wordpress/blocks';
-import { __ } from '@wordpress/i18n';
-import { IconButton } from '@wordpress/components';
 
 /**
  * Internal dependencies
  */
 import IgnoreNestedEvents from '../ignore-nested-events';
 import DefaultBlockAppender from '../default-block-appender';
-import Inserter from '../inserter';
+import ButtonBlockAppender from '../button-block-appender';
 
 function BlockListAppender( {
 	blockClientIds,
 	rootClientId,
 	canInsertDefaultBlock,
 	isLocked,
+	renderAppender,
 } ) {
 	if ( isLocked ) {
 		return null;
 	}
 
-	if ( canInsertDefaultBlock ) {
+	// If auto-insert Blocks is enabled, default to the standard behaviour
+	// of auto-inserting a Block but only if no renderAppender is provided.
+	if ( ! renderAppender && canInsertDefaultBlock ) {
 		return (
-			<IgnoreNestedEvents childHandledEvents={ [ 'onFocus', 'onClick', 'onKeyDown' ] }>
-				<DefaultBlockAppender
+			<div className="block-list-appender">
+				<IgnoreNestedEvents childHandledEvents={ [ 'onFocus', 'onClick', 'onKeyDown' ] }>
+					<DefaultBlockAppender
+						rootClientId={ rootClientId }
+						lastBlockClientId={ last( blockClientIds ) }
+					/>
+				</IgnoreNestedEvents>
+			</div>
+		);
+	}
+
+	// Render prop - custom appender.
+	if ( renderAppender ) {
+		return (
+			<div className="block-list-appender">
+				{ renderAppender() }
```

Comment:
On my prior point about being able to use the component as:

```jsx
<InnerBlocks renderAppender={ InnerBlocks.ButtonBlockAppender } />
```

...it depends on treating the render prop as a component definition. It probably works for plain function components, but also might be more correctly expressed as: `createElement( renderAppender )`.

I see most other resources which talk about render props ([[1]](https://cdb.reacttraining.com/use-a-render-prop-50de598f11ce) [[2]](https://reactjs.org/docs/render-props.html)) don't really mention this, and in most cases it doesn't make a difference, except if you'd want to reimplement `ButtonBlockAppender` as a class component.

Reply:
Interesting. 

`createElement( renderAppender )` would prevent any arguments being passed in the way render props typically do, though I suppose props could still be used.

I've left it as it is for now for consistency with other render props in the codebase, but could be something to explore in the future.

Reply:
> `createElement( renderAppender )` would prevent any arguments being passed in the way render props typically do, though I suppose props could still be used.

This is the way I tend to see render props used anyways (passed a props object, if anything), so it confused me that there's a tendency to treat them as components by convention, but not through actual use of `createElement`.